pkg/controllers/node/termination/terminator/terminator.go

```diff
@@ -115,18 +130,62 @@ func (t *Terminator) Evict(pods []*v1.Pod) {
 			}
 		}
 	}
-	// 2. Evict in order:
-	// a. non-critical non-daemonsets
-	// b. non-critical daemonsets
-	// c. critical non-daemonsets
-	// d. critical daemonsets
-	if len(nonCriticalNonDaemon) != 0 {
-		t.evictionQueue.Add(nonCriticalNonDaemon...)
-	} else if len(nonCriticalDaemon) != 0 {
-		t.evictionQueue.Add(nonCriticalDaemon...)
-	} else if len(criticalNonDaemon) != 0 {
-		t.evictionQueue.Add(criticalNonDaemon...)
-	} else if len(criticalDaemon) != 0 {
-		t.evictionQueue.Add(criticalDaemon...)
+
+	// EvictInOrder evicts only the first list of pods which is not empty
+	// future Evict calls will catch later lists of pods that were not initially evicted
+	t.EvictInOrder(
+		nonCriticalNonDaemon,
+		nonCriticalDaemon,
+		criticalNonDaemon,
+		criticalDaemon,
+	)
+}
+
+func (t *Terminator) EvictInOrder(pods ...[]*v1.Pod) {
+	for _, podList := range pods {
+		if len(podList) > 0 {
+			// evict the first list of pods that is not empty, ignore the rest
+			t.evictionQueue.Add(podList...)
+			return
+		}
 	}
 }
+
+func (t *Terminator) DeleteExpiringPods(ctx context.Context, pods []*v1.Pod, nodeGracePeriodTerminationTime *time.Time) error {
+	for _, pod := range pods {
+		// check if the node has an expiration time and the pod needs to be deleted
+		deleteTime := t.podDeleteTimeWithGracePeriod(nodeGracePeriodTerminationTime, pod)
+		if deleteTime != nil && time.Now().After(*deleteTime) {
+			// delete pod proactively to give as much of its terminationGracePeriodSeconds as possible for deletion
+			// ensure that we clamp the maximum pod terminationGracePeriodSeconds to the node's remaining expiration time in the delete command
+			gracePeriodSeconds := lo.ToPtr(int64(time.Until(*nodeGracePeriodTerminationTime).Seconds()))
+			t.recorder.Publish(terminatorevents.DisruptPodDelete(pod, gracePeriodSeconds, nodeGracePeriodTerminationTime))
+			opts := &client.DeleteOptions{
+				GracePeriodSeconds: gracePeriodSeconds,
+			}
+			if err := t.kubeClient.Delete(ctx, pod, opts); err != nil && !apierrors.IsNotFound(err) { // ignore 404, not a problem
+				return err // otherwise, bubble up the error
+			}
+			log.FromContext(ctx).WithValues(
+				"namespace", pod.Namespace,
+				"name", pod.Name,
+				"pod.terminationGracePeriodSeconds", *pod.Spec.TerminationGracePeriodSeconds,
+				"delete.gracePeriodSeconds", *gracePeriodSeconds,
+				"nodeclaim.expirationTime", *nodeGracePeriodTerminationTime,
+			).Info("deleting pod")
```

Comment:
I think this will only be noisy in scenarios where cluster admins have poorly configured terminationGracePeriod values compared to the terminationGracePeriodSeconds of their users' pods. When using this feature well, I expect that pods should be terminated gracefully via eviction API rather than pod deletes.

I would prefer to keep it as a log to make it more visible to cluster admins when pods are being frequently deleted instead of gracefully evicted. This information could be found by checking pod events, but not every cluster has a way to save events for later inspection.

If users find it problematic I'm happy to change it to debug level without much fuss.

Reply:
After re-reading the code, I think that even if the TGP and TGPS are configured properly, it's more about if the `do-not-disrupt` or PDB blocking pods don't get cleaned up by the nodepool.TGP - pod.TPGS, then we'll hit this log line. I imagine since we're changing this to eventual, this occurrence will be a bit more frequent. 

IIUC, this log line is hit even if nodepoolTGP > pod.TPGS.

Reply:
The log is only output when a pod is deleted. A pod is only deleted when T = nodepool.TGP - pod.TPGS.

- The do-not-disrupt annotation should be limited to use on ephemeral pods that should handle cleaning themselves up prior to the node termination time being reached, avoiding the pod delete. I realize a lot of karpenter users may be abusing this annotation for non-ephemeral pods (such as Deployments) and that will consistently result in those pods being deleted with a log line.
- PDB blocking pods should gracefully terminate long before the node's configured TGP. Pod deletes and logs for this case should only occur in situations where PDBs are blocked unexpectedly and other unusual circumstances.

Reply:
Chatted with @njtran about this over DM. We're in agreement that identifying cases where Karpenter is explicitly deleting pods rather than allowing them to gracefully terminate is important to users, but that identifying these cases may be better handled with events and metrics rather than logs.

Because events are ephemeral and historical metric aggregation may not always be reliable, I think it's a reasonable compromise to move this log to debug level in case cluster operations need a way to expose it through logging.